src/d3a/d3a_core/sim_results/stats.py

```diff
@@ -51,39 +51,56 @@ def total_avg_trade_price(markets):
     )
 
 
-def energy_bills(area, past_market_types, from_slot=None, to_slot=None):
+def _store_bought_trade(result_dict, trade_offer):
+    result_dict['bought'] += trade_offer.energy
+    result_dict['spent'] += trade_offer.price
+    result_dict['total_energy'] += trade_offer.energy
+    result_dict['total_cost'] += trade_offer.price
+
+
+def _store_sold_trade(result_dict, trade_offer):
+    result_dict['sold'] += trade_offer.energy
+    result_dict['earned'] += trade_offer.price
+    result_dict['total_energy'] -= trade_offer.energy
+    result_dict['total_cost'] -= trade_offer.price
+
+
+def energy_bills(area, past_market_types, external_trades, from_slot=None, to_slot=None):
     """
     Return a bill for each of area's children with total energy bought
     and sold (in kWh) and total money earned and spent (in cents).
     Compute bills recursively for children of children etc.
     """
     if not area.children:
-        return None
+        return None, None
     result = {child.name: dict(bought=0.0, sold=0.0,
                                spent=0.0, earned=0.0,
                                total_energy=0, total_cost=0,
                                type=get_area_type_string(child))
               for child in area.children}
+    external_trades[area.name] = dict(bought=0.0, sold=0.0,
+                                      spent=0.0, earned=0.0,
+                                      total_energy=0, total_cost=0)
```

Comment:
When I run the simulation, the "External Trade" member of the bills for each non-device area has a "type" member. Where is this set? I dont see it in the code. Can you help me please?

Reply:
The reason for that is this line https://github.com/gridsingularity/d3a/blob/feature/D3ASIM-1007/src/d3a/d3a_core/sim_results/endpoint_buffer.py#L190. External trades are actually generated from the flattened area results, since we only need to care about the parent area results. Thanks for pointing that out though, because I noticed that the external_trades are no longer used in the code. I will consequently remove these.